### ISSUE TITLE
ConnectionException: Attempt to connect to Node timed out after 60000ms. Process exited: True. Exit status: Process has been disposed.

### ISSUE THREAD

**Issue by aKzenT** [JeringTech/Javascript.NodeJS#143]
We had a partial outtage on our production website yesterday with all requests to node failing with:

`Attempt to connect to Node timed out after 60000ms. Process exited: True. Exit status: Process has been disposed.`

Before this we get a warning with the same text:
![image](https://user-images.githubusercontent.com/2589184/185643666-150df572-a2f5-49f2-8189-80cb84aea01f.png)

This came up directly after a configuration change on the azure web app, which caused the app to restart. What I don't really get is why it was not able to recover from this error automatically. If the process has exited / is disposed, shouldn't it just create a new process?

Also should the 60s timeout really apply if the process itself has exited or could that also be improved, so that it waits on either being able to connect or the process being killed for whatever reason?

**Comment by aKzenT** [comment]
It could be helpful to log the process id of the affected process in the error message as well. This would help to see if the process id is always the same or if there have been in fact various processen involved.

**Comment by JeremyTCD** [comment]
Apologies for issues.

Could you paste the full log message for the first warning? The exception that caused the connection attempt to fail is appended at the end.

To answer your questions:

> If the process has exited / is disposed, shouldn't it just create a new process?

Yes it does. After a connection attempt fails, a new process is created for the next attempt. So your app tried to connect to one process, failed, created a new one and timed out on the new one.

> so that it waits on either being able to connect or the process being killed for whatever reason?

It waits on being able to connect. To my knowledge there is no easy way to know when the NodeJS process dies unexpectedly. So even if it dies, the .Net process waits till the timeout is over before trying to connect again (if retries remain).

I think perhaps the default of 60s is too long. Successful connections generally take << 1s. Also the default NumConnectionRetries should be -1 (infinite) since failure to connect is terminal applications. 

A timeout of 3s and NumConnectionRetries of -1 would probably have prevented the disruption to your service.


 You can change the timeout by setting [OutOfProcessNodeJSServiceOptions.TimeoutMS](https://github.com/JeringTech/Javascript.NodeJS#outofprocessnodejsserviceoptionstimeoutms).

**Comment by JeremyTCD** [comment]
Apologies for issues.

Could you paste the full log message for the first warning? The exception that caused the connection attempt to fail is appended at the end.

To answer your questions:

> If the process has exited / is disposed, shouldn't it just create a new process?

Yes it does. After a connection attempt fails, a new process is created for the next attempt. So your app tried to connect to one process, failed, created a new one and timed out on the new one.

> so that it waits on either being able to connect or the process being killed for whatever reason?

It waits on being able to connect. To my knowledge there is no easy way to know when the NodeJS process dies unexpectedly. So even if it dies, the .Net process waits till the timeout is over before trying to connect again (if retries remain).

I think perhaps the default of 60s is too long. Successful connections generally take << 1s. Also the default NumConnectionRetries should be -1 (infinite) since failure to connect is terminal for most applications. 

A timeout of 3s and NumConnectionRetries of -1 would probably have prevented the disruption to your service. Thoughts on making these the defaults?

For the time being you can manually set [OutOfProcessNodeJSServiceOptions.TimeoutMS](https://github.com/JeringTech/Javascript.NodeJS#outofprocessnodejsserviceoptionstimeoutms) and [OutOfProcessNodeJSServiceOptions.NumConnectionRetries](https://github.com/JeringTech/Javascript.NodeJS#outofprocessnodejsserviceoptionsnumconnectionretries).

**Comment by JeremyTCD** [comment]
As for:

> It could be helpful to log the process id of the affected process in the error message as well.

and 

> Process exited: True. Exit status: Process has been disposed." information is also a bit misleading.

I agree with both points. The exception message should be created before disposing so we get the exit code. I will implement both changes.

**Comment by aKzenT** [comment]
Hey,

no worries. Thank you for this great library.

The following is the complete message I have for the warning:
```
An attempt to connect to NodeJS failed. Retries remaining: 1.
Exception:
  Jering.Javascript.NodeJS.ConnectionException: Attempt to connect to Node timed out after 60000ms. Process exited: True. Exit status: Process has been disposed.
   at Jering.Javascript.NodeJS.OutOfProcessNodeJSService.ConnectIfNotConnected()
```

Reducing the timeout to 3s is actually what we did now as well as we came to the same conclusion that 60s is very long. Especially in a web server situation where requests will rapidly pile up. If you account for the default retries you will even wait 2min in total. So I'm in favour of changing the default here.

I believe you can use the process "Handle" property to wait for process exits. See here for an example which is very similar to your situation: https://stackoverflow.com/a/26852562

**Comment by JeremyTCD** [comment]
No problem, hope the disruption wasn't too severe.

Agreed on the defaults, I'd recommend increasing connection retries too. Will implement all of the changes we've discussed since there are no glaring cons to doing so. Let's leave this issue open till the changes are merged.

**Comment by aKzenT** [comment]
I'm not sure if increasing the connection retries is a good idea. If it failed two times in a row it's unlikely a third time will succeed. If we wait 3s for each try, we are already at 9s for doing 3 tries. I would argue that for many websites failing fast is more important than retrying a long time as the latter can easily lead to a server filling up with pending requests which can bring down the server completely. Also retrying too often by default could hide fundamental issues in the application itself which should better be fixed instead of retried until it works.

Having one retry and short timeouts seems like a fair compromise that should provide high stability and fast failure.

Did you take a look at the WaitHandle stuff? I think that would have the biggest impact for us since we could just skip the whole wait times when the process already exited anyway.

**Comment by JeremyTCD** [comment]
Noted on retries, was thinking the same thing so did some investigating. 

I've reproduced connection timeouts on my machine, turns out the issue is thread pool starvation. `ThreadPool.SetMinThreads()` is a workaround till the code is improved.

Some details: 

`OutOfProcessNodeJSService.ConnectIfNotConnected` blocks all threads other than the one starting the Node.js process. This ties up a bunch of threads, so the runtime has to create new threads for other tasks.

Once the runtime has created `ThreadPool.GetMinThreads()` number of threads, which defaults to number-of-cores and is fairly low for must Azure Web App instances, thread creation is throttled to 1-2 per second. 

Because of throttling, it takes a long time for a thread to be available to read `stdout` from the Node.js process. Reading `stdout` is necessary to establish the connection.

To fix the issue I'm converting the locks in `OutOfProcessNodeJSService` to async locks using `SemaphoreSlim`. Wrt `WaitHandle`, I ended up using `Process.Exited` instead.

Will publish a new version next week.

**Comment by aKzenT** [comment]
That sounds great! Looking forward to the new version.

**Comment by JeremyTCD** [comment]
I made some the changes we discussed in this issue:

> - Split OutOfProcessNodeJSService.TimeoutMS into OutOfProcessNodeJSService.InvocationTimeoutMS and OutOfProcessNodeJSService.ConnectionTimeoutMS. Why? TimeoutMS was the maximum duration for connecting to NodeJS and for invocations. Since invocations can take a while, it had a default of 60,000. This default was not ideal for
connecting to NodeJS - if something went wrong with the target NodeJS process, it took a while for the connection attempt to timeout. The split solves this dilemma.
> - Added mechanism that immediately terminates a connection attempt when the target process exits before connecting.

Full details: [Improve process connection system](https://github.com/JeringTech/Javascript.NodeJS/pull/144#issue-1345085475). Changes available in 7.0.0-beta.1.

Thanks for the suggestions and thanks again for the contributions, feel free to open a new issue if you encounter any.

**Comment by aKzenT** [comment]
@JeremyTCD I installed the latest beta today and so far everything seems to work. I also tested and confirmed that the issue of the additional headers described here is fixed by that release. Thanks!
